File: frontend/src/api/contratos/denuncia.ts
import * as z from 'zod';

export const tipoAlvoDenunciaSchema = z.enum(['SERVICO', 'PERFIL', 'AVALIACAO']);

export type TipoAlvoDenuncia = z.infer<typeof tipoAlvoDenunciaSchema>;

export const statusDenunciaSchema = z.enum(['PENDENTE', 'PROCEDENTE', 'IMPROCEDENTE']);

export type StatusDenuncia = z.infer<typeof statusDenunciaSchema>;

export const denunciaSchema = z.object({
  id: z.uuid(),
  autorPerfilId: z.uuid(),
  denunciadoPerfilId: z.uuid().nullable(),
  contratacaoId: z.uuid().nullable(),
  alvoTipo: tipoAlvoDenunciaSchema,
  alvoId: z.string(),
  motivo: z.string(),
  descricao: z.string().nullable(),
  status: statusDenunciaSchema,
  analisadoPorPerfilId: z.uuid().nullable(),
  analisadoEm: z.string().nullable(),
  parecer: z.string().nullable(),
  criadoEm: z.string(),
  atualizadoEm: z.string(),
});

export type Denuncia = z.infer<typeof denunciaSchema>;

export const denunciaRequestSchema = z.object({
  alvoTipo: tipoAlvoDenunciaSchema,
  alvoId: z.uuid(),
  motivo: z
    .string({ error: 'O motivo é obrigatório' })
    .min(1, 'O motivo é obrigatório')
    .max(80, 'Máximo de 80 caracteres'),
  descricao: z.string().max(1000, 'Máximo de 1000 caracteres').optional(),
  denunciadoPerfilId: z.uuid().optional(),
  contratacaoId: z.uuid().optional(),
});

export type DenunciaRequest = z.infer<typeof denunciaRequestSchema>;

File: backend/src/main/java/br/com/bicoemcasa/api/modulos/avaliacoes/models/Denuncia.java
package br.com.bicoemcasa.api.modulos.avaliacoes.models;

import jakarta.persistence.*;
import lombok.AccessLevel;
import lombok.AllArgsConstructor;
import lombok.Builder;
import lombok.Getter;
import lombok.NoArgsConstructor;
import lombok.Setter;

import java.time.OffsetDateTime;
import java.util.UUID;
import java.util.List;

import org.hibernate.annotations.JdbcTypeCode;
import org.hibernate.type.SqlTypes;

@Entity
@Table(name = "denuncia")
@Getter
@Setter
@Builder
@NoArgsConstructor(access = AccessLevel.PROTECTED)
@AllArgsConstructor

public class Denuncia {
    @Id 
    @GeneratedValue(strategy = GenerationType.UUID)
    private UUID id ; 

    @Column(name = "autor_perfil_id" , nullable = false)
    private  UUID  autorPerfilId;

    @Column(name = "denunciado_perfil_id")
    private  UUID  denunciadoPerfilId;

    @Column(name = "contratacao_id")
    private UUID contratacaoId;

    @Enumerated(EnumType.STRING)
    @JdbcTypeCode(SqlTypes.NAMED_ENUM)
    @Column(name="alvo_tipo" , nullable = false , columnDefinition = "tipo_alvo_denuncia")
    private TipoAlvoDenuncia alvoTipo; 

    @Column(name="alvo_id" , nullable = false,length = 64)
    private String alvoId;

    @Column(nullable = false,length = 80)
    private String motivo;

    @Column(length = 1000)
    private String descricao;

    @JdbcTypeCode(SqlTypes.JSON)
    private List<String> fotos;

    @Enumerated(EnumType.STRING)
    @JdbcTypeCode(SqlTypes.NAMED_ENUM)
    @Column(nullable = false , columnDefinition = "status_denuncia" )
    private StatusDenuncia status;

    @Column(name="analisado_por_perfil_id" )
    private UUID  analisadoPorPerfilId;

    @Column(name="analisado_em" )
    private OffsetDateTime analisadoEm;

    @Column(length = 1000)
    private String parecer;

    @Column(name = "criado_em" , nullable = false , updatable = false)
    private OffsetDateTime criadoEm;

    @Column(name = "atualizado_em" , nullable = false)
    private OffsetDateTime atualizadoEm;


    @PrePersist 
    protected void criar(){
        OffsetDateTime agora = OffsetDateTime.now() ; 
        this.criadoEm = agora ; 
        this.atualizadoEm=agora;
        this.status = StatusDenuncia.PENDENTE;
    }

    @PreUpdate 
    protected void atualizar(){
        this.atualizadoEm= OffsetDateTime.now();
    }

}
